File: 09-backtracking/00131-Palindrome_Partitioning/00131-palindrome_partitioning.cpp
// 131: Palindrome Partitioning
// https://leetcode.com/problems/palindrome-partitioning



#include <iostream>
#include <vector>

using namespace std;

class Solution {
private:
    void backtrack(string s, int start, vector<string>& current, vector<vector<string>>& result) {
        if (start == s.size()) {
            result.push_back(current);
            return;
        }

        for (int i=start; i<s.size(); i++) {
            if (isPalindrome(s, start, i)) {
                string str = s.substr(start, i - start + 1);
                current.push_back(str);
                backtrack(s, i + 1, current, result);
                current.pop_back();
            }
        }
    }

    bool isPalindrome(string s, int left, int right) {
        while (left < right) {
            if (s[left] != s[right]) return false;
            left++;
            right--;
        }

        return true;
    }

public:
    // SOLUTOIN
    vector<vector<string>> partition(string s) {
        vector<string> current;
        vector<vector<string>> result;
        backtrack(s, 0, current, result);
        return result;
    }
};


int main() {
    Solution o;

    // INPUT
    string s = "aab";

    // OUTPUT
    auto result = o.partition(s);

    cout << "["; for (auto x : result) {
        cout << "["; for (auto y : x) {
            cout << y << ",";
        } cout << "\b],";
    } cout << "\b]" << endl;

    return 0;
}

File: 09-backtracking/00131-Palindrome_Partitioning/00131-palindrome_partitioning.java
// 131: Palindrome Partitioning
// https://leetcode.com/problems/palindrome-partitioning


import java.util.List;
import java.util.ArrayList;

class Solution {
    private void backtrack(String s, int start, List<String> current, List<List<String>> result) {
        if (start == s.length()) {
            result.add(new ArrayList<>(current));
            return;
        }

        for (int i=start; i<s.length(); i++) {
            if (isPalindrome(s, start, i)) {
                String str = s.substring(start, i + 1);
                current.add(str);
                backtrack(s, i + 1, current, result);
                current.remove(current.size() - 1);
            }
        }
    }

    private boolean isPalindrome(String s, int left, int right) {
        while (left < right) {
            if (s.charAt(left) != s.charAt(right)) return false;
            left++;
            right--;
        }

        return true;
    }

    
    // SOLUTION
    public List<List<String>> partition(String s) {
        List<List<String>> result = new ArrayList<>();
        List<String> current = new ArrayList<>();
        backtrack(s, 0, current, result);
        return result;
    }

    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        String s = "aab";

        // OUTPUT
        var result = o.partition(s);
        System.out.println(result);
    }
}
